src/Mod/Sketcher/Gui/ViewProviderSketch.h

```diff
@@ -139,7 +139,7 @@ using GeoListFacade = Sketcher::GeoListFacade;
  * concentrating the coupling in a single point (and code reuse).
  *
  */
-class SketcherGuiExport ViewProviderSketch : public PartGui::ViewProvider2DObjectGrid
+class SketcherGuiExport ViewProviderSketch : public PartGui::ViewProvider2DObject

```

Comment:
Just dropping by to check if some questions needs answers.

Why do you say it's not acceptable to have the grid properties in the VPSketch? There are other properties there. It make 10 times more sense to have them here rather than in a parent class hidden.

Having the VPSketch inherinting from a Part class ViewProvider2DObjectGrid was beyond unintuitive. It just feels plain hacky wrong.
I have done exactly what I have been asked to. Leave ViewProvider2DObjectGrid alone for "potential external users" of this object and implemented in a coinmanager. You're even the one who said it would be best as a CoinManager after OpenBrain said to move it inside VPSketch!

Before I had just overriden ViewProvider2DObjectGrid, and @0penBrain told me to leave it alone because it's a shared object that maybe is used somewhere else. So I moved to VPSketch, then it has been asked to move to CoinManager. So I complied.

Now you're telling me it's not acceptable and it needs to do what I have done initially. Or making things even more unecessarilly complex than they were by making a subclass of ViewProvider2DObjectGrid and inherinting from it.

This is a mad house.

Reply:
> I have done exactly what I have been asked to. Leave ViewProvider2DObjectGrid alone for "potential external users" of this object and implemented in a coinmanager. You're even the one who said it would be best as a CoinManager after OpenBrain said to move it inside VPSketch!

Sorry I can't jump in into this before. @abdullahtahiriyo yes @PaddleStroke is true. This was my assessment of the situation that the best solution overall was to keep VP2DObjectGrid untouched ATM and provide the new implementation into VPSketch. The main argument for not inheriting VP2DObjectGrid was that we imagine to drop properties (at least `TightGrid` IIRC) that would have been forcefully brought in case of inheritance. If you think we can't live with this architecture, I will provide the implementation to change it to the needs.

> Having the VPSketch inherinting from a Part class ViewProvider2DObjectGrid was beyond unintuitive. It just feels plain hacky wrong.
I think I already told that such statement tends to feel like you're missing some basic concepts of object-oriented programming. This absolutely was not wrong.
Also I may only encourage you to clarify and improve your communication. IIRC you opened a dedicated thread on the forum about the grid refactoring, and this thread is even not mentioned in the PR description. ;)

Reply:
> This was my assessment of the situation that the best solution overall was to keep VP2DObjectGrid untouched ATM and provide the new implementation into VPSketch. The main argument for not inheriting VP2DObjectGrid was that we imagine to drop properties (at least `TightGrid` IIRC) that would have been forcefully brought in case of inheritance. If you think we can't live with this architecture, I will provide the implementation to change it to the needs.

By no means. 

I seek to understand where we are coming from and where we are going to, as the decisions on architecture need to take into account the requirements.

If we want to part ways with VP2dObjectGrid for technical reasons (properties is a very valid one), sure we need to implement the properties elsewhere. This 'elsewhere' may be in a new parent class of VPSketch taking care of the grid, or in VPSketch itself.

As code reuse apparently not foreseable, as the grid requirements appear very specific, the benefits of a new parent class are merely separating code and contributing to keep VPSketch growing under control. In what Paddle fails to see as an advantage, is that a developer wanting to extend the behaviour of the grid, will only need to focus on a very small class, where all grid code would be contained. Somehow, there is an encapsulation of the code dealing with the grid and a degree of functional decoupling from the rest of the very extensive ViewProviderSketch (in other words, one can forget there is a grid involved when looking at VPSketch). 

Bringing the properties and all the code to VPSketch does allow for a stronger coupling between the grid and the rest of the VPSketch code. If such stronger coupling is wanted (because there in strong interaction between grid and the rest of the VP), then we should look for a minimal impact on the ViewProvider itself, and rely on a delegate (a coin manager alike class) for all coin management. Such solution would have a bigger blueprint on VPSketch, but it would still keep the lion's share of the growing under control. I understand this is what you had in mind.

Reply:
> Bringing the properties and all the code to VPSketch does allow for a stronger coupling between the grid and the rest of the VPSketch code. If such stronger coupling is wanted (because there in strong interaction between grid and the rest of the VP), then we should look for a minimal impact on the ViewProvider itself, and rely on a delegate (a coin manager alike class) for all coin management.

This is exactly the target, and normally what should be now in the PR (maybe you were too busy to review it recently).
The advantage of having the code in VPSketch is that it can benefit directly from things like the camera sensor I implemented earlier. If VPSketch becomes too big, we could push this to a new base class, but I think this goes out of scope of this PR.
On the opposite, I "enforced" @PaddleStroke to move a maximum part of the code (especially everything related to Coin) into a new CoinManager delegate class. What he successfully did.

At this time I think we all need to let this cool down a bit. I will do another complete review when I can.

Discussions purely about architecture are still very welcome. Again, if some decisions should go against my original ones, I'm very fine with this and will help providing the needed rework.

Reply:
Ok then for local properties + coinmanager.

If after your prospective new complete review, you want me to look into it, just let me know.

Thank you.

Reply:
> would you be so kind to give me your honest opinion about this solution with the inheritance?

The most flexible design is to:
* implement a class ViewProviderGridExtension that inherits from ViewProviderExtension
* the class ViewProviderGridExtension then will use the methods of EditModeGridCoinManager
* a view provider that needs "grid behaviour" either additionally inherits from  ViewProviderGridExtension or uses it as a class member

Because a grid is not only relevant for sketching but can be useful for Draft the two classes should go to PartGui.

Currently ViewProvider2DObjectGrid is only used as sub-class of ViewProviderSketch and isn't used anywhere else. So, I don't know why it shouldn't be touched to use the new grid stuff. But if there is any real need to keep the old grid stuff then an option is to move all this logic to another view provider extension class. ViewProvider2DObjectGrid then uses ViewProviderGridExtension and ViewProviderSketch still inherits from ViewProvider2DObjectGrid.

Doing the grid handling inside ViewProviderSketch is a sub-optimal solution because it violates the Single Responsibility Principle and additionally makes it almost impossible to re-use existing code.

Reply:
@wwmayer 
Now that I am fixing this part, I wanted to say thank you.